crates/primitives/src/env.rs

```diff
@@ -458,11 +460,40 @@ impl Env {
         // - For CANCUN and later, check that the gas price is not more than the tx max
         // - For before CANCUN, check that `blob_hashes` and `max_fee_per_blob_gas` are empty / not set
         if SPEC::enabled(SpecId::CANCUN) {
+            // Presence of max_fee_per_blob_gas means that this is blob transaction.
             if let Some(max) = self.tx.max_fee_per_blob_gas {
+                // ensure that the user was willing to at least pay the current blob gasprice
                 let price = self.block.get_blob_gasprice().expect("already checked");
                 if U256::from(price) > max {
                     return Err(InvalidTransaction::BlobGasPriceGreaterThanMax);
                 }
+
+                // there must be at least one blob
+                // assert len(tx.blob_versioned_hashes) > 0
+                if self.tx.blob_hashes.is_empty() {
+                    return Err(InvalidTransaction::EmptyBlobs);
+                }
+
+                // The field `to` deviates slightly from the semantics with the exception
+                // that it MUST NOT be nil and therefore must always represent
+                // a 20-byte address. This means that blob transactions cannot
+                // have the form of a create transaction.
+                if self.tx.transact_to.is_create() {
+                    return Err(InvalidTransaction::BlobCreateTransaction);
+                }
+
+                // all versioned blob hashes must start with VERSIONED_HASH_VERSION_KZG
+                for blob in self.tx.blob_hashes.iter() {
+                    if blob[0] != VERSIONED_HASH_VERSION_KZG {
+                        return Err(InvalidTransaction::BlobVersionedHashesNotSupported);
```

Comment:
it is different error, one is related to precompile input and second one is related to transaction blobs.

Reply:
yeah, I'm mainly just suggesting a separate error variant similar to the precompile one. In reth, especially in rpc responses, it would be nice to distinguish between `BlobVersionedHashesNotSupported` and another error type which indicates that a versioned hash has an invalid version